src/theory/quantifiers/ceg_t_instantiator.cpp

```diff
@@ -1180,10 +1165,34 @@ Node BvInstantiator::rewriteAssertionForSolvePv( Node pv, Node lit ) {
     it = visited.find(cur);
 
     if (it == visited.end()) {
-      visited[cur] = Node::null();
-      visit.push(cur);
-      for (unsigned i = 0; i < cur.getNumChildren(); i++) {
-        visit.push(cur[i]);
+      if (cur.getKind() == CHOICE)
+      {
+        // must replace variables of choice functions
+        // with new variables to avoid variable
+        // capture when considering substitutions
+        // with multiple literals.
+        Node bv = ci->getBoundVariable(cur[0][0].getType());
+        TNode var = cur[0][0];
+        Node sbody = cur[1].substitute(var, bv);
+        // we cannot cache the results of subterms
+        // of this choice expression since we are
+        // now in the context { cur[0][0] -> bv },
+        // hence we make a separate call to
+        // rewriteAssertionForSolvePv here,
+        // where the recursion depth is the maximum
+        // depth of nested choice expressions.
+        Node rsbody = rewriteAssertionForSolvePv(ci, pv, sbody);
```

Comment:
Even if the max recursion depth seems limited here, let's not introduce new recursive implementations and unrecursify rewriteAssertionForSolvePv (maybe in a separate PR).

Reply:
sounds good, it's possible to do this, just need to have two levels to the context in this function.